src/main/java/com/puppycrawl/tools/checkstyle/utils/TypeDeclDesc.java

```diff
@@ -0,0 +1,234 @@
+////////////////////////////////////////////////////////////////////////////////
+// checkstyle: Checks Java source code for adherence to a set of rules.
+// Copyright (C) 2001-2022 the original author or authors.
+//
+// This library is free software; you can redistribute it and/or
+// modify it under the terms of the GNU Lesser General Public
+// License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+//
+// This library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+// Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public
+// License along with this library; if not, write to the Free Software
+// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+////////////////////////////////////////////////////////////////////////////////
+
+package com.puppycrawl.tools.checkstyle.utils;
+
+import java.util.ArrayDeque;
+import java.util.Deque;
+
+import com.puppycrawl.tools.checkstyle.api.DetailAST;
+import com.puppycrawl.tools.checkstyle.api.TokenTypes;
+
+/**
+ * Maintains information about the type declaration.
+ * Any ast node of type {@link TokenTypes#CLASS_DEF} or {@link TokenTypes#INTERFACE_DEF}
+ * or {@link TokenTypes#ENUM_DEF} or {@link TokenTypes#ANNOTATION_DEF}
+ * or {@link TokenTypes#RECORD_DEF} is considered as a type declaration.
+ */
+public class TypeDeclDesc {
+
```

Comment:
This class does not look like a utility class. I see two ways:
Move this class to the package `com.puppycrawl.tools.checkstyle`,
or, create a base class for all checks that should use this class and put this class inside of it.

Reply:
> create a base class for all checks that should use this class and put this class inside of it.

This class is used by FinalClass and UnusedLocalVariable, creating a base class for them isn't the best idea as one governs the design and the other governs the coding aspect, though I am ok with the former solution.

Reply:
I agree with @pbludov 's idea to move this class out of utils, but I wonder if we should make it some abstract class, with each check having it's own subclass with specific implementation details. I suspect this will also help with mutation coverage, and leave details specific to each check out of main implementation.

@Vyom-Yadav right now, are both checks using all methods and fields in this class?

Reply:
> it some abstract class, with each check having it's own subclass with specific implementation details.

There is only one implementation of every method, though some classes do not use some of the methods.

> I suspect this will also help with mutation coverage 

Mutation coverage shouldn't be a problem with the current mutators
(+ REMOVE_CONDITIONALS), I tested it with REMOVE_CONDITIONALS as pitest-utils has that mutator commented out and every new line was covered.

> leave details specific to each check out of the main implementation.

This is a good idea, jumping from class to class to check out implementation can be tedious, but most of the fields are just boolean flags. Only `getUpdatedCopyOfVarStack(..)` has an implementation, `declaredAsFinal` and `declaredAsAbstract`are instantiated in the constructor.

> are both checks using all methods and fields in this class?

No only three fields are shared.

1. `qualifiedName`
2. `depth`
3. `typeDeclAst`

rest all the fields are particular to a check. (Some of them are not even applicable to the other check)

Reply:
@Vyom-Yadav if you think abstract class is a good idea, can you make a rough sketch of what this class would look like, and what subclasses might look like, as proposal? Nothing as fancy as lucidcharts is needed, just something in markdown that takes a few minutes.

Reply:
> Mutation coverage shouldn't be a problem

I should clarify; I meant that ideally, we would only have to add one class to pitest for abstract class, since one class would cover all abstract class code. Concrete class in check implementation would cover the rest in it's respective pitest.

Reply:
> This is a good idea, jumping from class to class to check out implementation can be tedious, **BUT** most of the fields are just boolean flags.
>> if you think abstract class is a good idea, can you make a rough sketch of what this class would look like

I meant to say usually it is a good idea but in this case, I feel having an abstract class for this might be overkill (We add the complexity of class hierarchy) but if I had to make an implementation it would be something like-
```java
public abstract class AbstractTypeDeclDesc {
    private final String qualifiedName;
    private final int depth;
    private final DetailAST typeDeclAst;

    public AbstractTypeDeclDesc(String qualifiedName, int depth,
                                DetailAST typeDeclAst) {
        this.qualifiedName = qualifiedName;
        this.depth = depth;
        this.typeDeclAst = typeDeclAst;
    }
}
```
Then `FinalClass` will have a subclass called `ClassDesc` which will extend this and the fields, getter/setters it requires. Similarly, `UnusedLocalVar` will also have a subclass extending it and adding its custom fields/methods.

Reply:
@nick-mancuso @pbludov Let's agree on a solution and move forward from there.

Reply:
If this is all they have in common, then I do not see the point in combining the two. What benefits would this provide, except to avoid duplicating ~12 lines of code? I am not sure this is a good enough reason to make two checks dependent on some other class.

Reply:
@Vyom-Yadav, this issue is more complicated than I thought. Perhaps it's better to leave these classes as they are for now.

Maybe we need to add some entities to the `api` package to represent types (as we have `FullIdent` for ids).
Something that makes it easier to work with nested classes.
We need to discuss this more carefully.

Reply:
Done.